### PR TITLE
Added basic support for adding subcaptions via :meth:`.Scene.add_subcaption`

### LINKED ISSUE #910: Feature request: Subtitle timings
## Description of proposed feature

It would be fairly useful to have animation options that aren't for actually displaying anything, but for marking the timestamps in an animation with text that can be used to export subtitles.

Since a lot of videos made in manim are intended to include additional voiceover, this would make timing subtitles, or at least adding subtitles for the intended script to videos, a lot easier. Even if the actual voiceover doesn't 100% match the script, it's a lot easier to tweak the words and the timing than to make them from scratch.

Alternatively, folks who don't do voiceover but want to add subtitles as an alternative can still do so.

## How can the new feature be used?

I'm imagining an implementation similar to:

1. Animations can set the subtitle text or clear it (e.g. end the latest subtitle), which would happen as soon as the animation starts.
2. There is a dedicated animation `Subtitle(text)` which sets the subtitle to the given string, or clears it if passed `None`.
3. Alternatively, there could be dedicated `set_subtitle` and `clear_subtitle` methods since setting a subtitle takes zero time.

This way, you can still use the existing animation system to leverage adding subtitles, e.g. `wait` still works as expected and adding a subtitle animation alongside other animations would set it as soon as those animations are started.  

## Additional comments

The output would probably make the most sense as an srt file since the videos are output as MP4. This is a relatively simple format where each line includes a start/end timestamp for a subtitle and the text. The lines could be written out whenever a subtitle is cleared or changed, or the video ends.

### PR DESCRIPTION
**Pull request by behackl**
## Overview: What does this pull request change?

- New method `Scene.add_subcaption`
- New keyword arguments `subcaption`, `subcaption_duration`, `subcaption_offset` for `Scene.play`

## Motivation and Explanation: Why and how do your changes improve the library?

If a user writes any subcaptions in a Scene via the newly introduced methods (or kwargs), an `.srt` file is written containing these subcaptions.

Closes #910.

## Links to added or changed documentation pages

Documentation of Scene: https://manimce--2314.org.readthedocs.build/en/2314/reference/manim.scene.scene.Scene.html#manim.scene.scene.Scene

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by christopher-besch** [comment]
I'm a bit late to the game but I've already thought about such a feature. With the difference that I would have defined the captions in the sections, and not the play commands. That way the captions could also be displayed in the Manim Editor, making it easier to record the audio and rehearse a script.

This should still work if you're not rendering with section video output.

What do you say?

**Comment by behackl** [comment]
> That way the captions could also be displayed in the Manim Editor, making it easier to record the audio and rehearse a script.

In order to use the subcaptions for the Manim Editor, it should be sufficient to store them in the sections instead of in an attribute of the file writer, no? I do find the current interface where the subcaptions can be specified in the play calls rather useful, so I would like to keep that -- but I don't see any harm in making this feature more compatible with the section API.

My suggestion would be:
- I'll split `Scene.renderer.file_writer.subcaptions` up and store the subcaption in the latest section (this might not always be semantically correct though: what if someone specifies a large time offset that moves the caption out of the section that is currently playing to a previous one?),
- And then I'll add a property, `Scene.renderer.file_writer.subcaptions` which again outputs a list of all subcaptions by combining the subcaptions stored in all sections into one list again.

**Comment by christopher-besch** [comment]
> [...]what if someone specifies a large time offset that moves the caption out of the section that is currently playing to a previous one?

My idea was that the user doesn't even have to define any duration, it would simply use the length of the current section for that.
I think it's kinda difficult to keep track of how much time has passed so I'd be convenient for Manim to do that.
But if a users can't live without custom durations, that could be made possible—at the risk of breaching a section boundary.
Then the Manim Editor would have to handle those breaches...somehow.

> [...]combining the subcaptions[...]

I definitely don't want to give up the `.srt` files and this step seems necessary for that to work.

**Comment by behackl** [comment]
I have thought some more about changing the implementation, and have changed my mind about storing the subtitles in the section. Given that users might change the time stamp when and for how long subcaptions are displayed, this information might diverge substantially from the sections.

For the Manim Editor, I think it would make more sense if you added support for .srt files – then, you could simply display some markers in the timeline where the subcaptions are up. Also, it would allow people to use their own subcaption files, and not depend on anything Manim-internal.

What do you think? :-)

**Comment by christopher-besch** [comment]
> diverge substantially from the sections

I'm not so sure if this is the case and am not too happy with throwing even more complexity at the user. Now there are scenes, sections, animations, frames, subcaptions and whatnot. I've already received some criticism that this is quite a lot for beginners so I would've been happy to combine at least subcaptions and sections.
But I do get your point; so I'll implement the support for the `.srt` files.

> simply display some markers in the timeline

You don't even know how painful adding new UI elements is xD
I'll see what I can do.

**Comment by behackl** [comment]
> I've already received some criticism that this is quite a lot for beginners

True, there are many different concepts that are in use -- and while that is certainly overwhelming for completely new users, it adds a lot of flexibility to more experienced ones.

I feel like the use case where one section would have more than one subcaption (e.g., if you have a long animation where the text doesn't fit in just one caption) is not that exotic, but becomes difficult to handle if subcaption display durations are simply set as the section durations. That's why I think they should be handled separately.

> You don't even know how painful adding new UI elements is xD

I've done my fair share of UI and web stuff, I definitely feel you. My comment wasn't meant in the sense that it would be easy to implement, but rather in the sense that there should be some "simple" indication of where (and when) subcaptions happen in your file. I know that handling that is a lot of work – thank you for your efforts in that regard. :-)

File: manim/scene/scene_file_writer.py [hunk 1]
```diff
@@ -712,6 +716,13 @@ def flush_cache_directory(self):
             {"par_dir": self.partial_movie_directory},
         )
 
+    def write_subcaption_file(self):
+        """Writes the subcaption file."""
+        subcaption_file = Path(config.output_file).with_suffix(".srt")
+        with open(subcaption_file, "w") as f:
+            f.write(srt.compose(self.subcaptions))
+        logger.info(f"Subcaption file has been written as {subcaption_file}")
```
Comment (christopher-besch) [inline manim/scene/scene_file_writer.py:724]: This function is supposed to stay untested, right?

File: manim/scene/scene.py [hunk 2]
```diff
@@ -899,8 +902,46 @@ def get_run_time(self, animations):
...
+        start_time = self.renderer.time
         self.renderer.play(self, *args, **kwargs)
+        run_time = self.renderer.time - start_time
+        if subcaption:
+            if subcaption_duration is None:
+                subcaption_duration = run_time
+            self.add_subcaption(
+                content=subcaption,
+                duration=subcaption_duration,
+                offset=-run_time + subcaption_offset,
```
Comment (christopher-besch) [inline manim/scene/scene.py:943]: Why are you subtracting the `run_time` here? And if this is intended, we have two different `offset`s, which is rather confusing, isn't it?

**Review by christopher-besch — CHANGES_REQUESTED** [review]

File: manim/scene/scene_file_writer.py (hunk 1 above)
Comment (behackl) [reply manim/scene/scene_file_writer.py:724]: I think this is fine, yes. There is nothing exotic going on here, and the more critical part is the creation of the subcaptions.

File: manim/scene/scene.py (hunk 2 above)
Comment (behackl) [reply manim/scene/scene.py:943]: It is intended: the subcaption is added after the animation has been played, which means that the time stamp (`self.renderer.time`) has already been updated to the time after the animation has been played. To get the subcaption to be displayed when the animation starts playing, its starting time needs to be shifted by the run time of the animation.

What do you mean with two different `offset`s?

Comment (christopher-besch) [reply manim/scene/scene.py:943]: > What do you mean with two different `offsets`?

We have both a `subcaption_offset` and an `offset` here. I know this is quite picky but it still confused me.

Comment (behackl) [reply manim/scene/scene.py:943]: Ah – I see! I think it makes sense for the keyword argument for `Scene.play` to be called `subcaption_offset`, to explicitly make clear that we are talking about the offset of subcaptions there -- but when using `Scene.add_subcaption`, naming the corresponding parameter `subcaption_offset` is not really necessary I believe.

I could live with refactoring it, if you'd strongly prefer that. :-)

**Review by christopher-besch — APPROVED** [review]

**Pull request merged**